### PR TITLE
8280703: CipherCore.doFinal(...) causes potentially massive byte[] allocations during decryption

### PR DESCRIPTION
**Pull request by asgibbons**
Backporting because this change results in ~3x performance improvement in AES-CTR.

Risk is low.  Tested with tier1 and benchmark.

The PR does not backport cleanly.  The buffer name was changed, but functionally remains the same.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8280703](https://bugs.openjdk.org/browse/JDK-8280703): CipherCore.doFinal(...) causes potentially massive byte[] allocations during decryption

### Reviewers
 * [Christoph Langer](https://openjdk.org/census#clanger) (@RealCLanger - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk11u-dev pull/1780/head:pull/1780` \
`$ git checkout pull/1780`

Update a local copy of the PR: \
`$ git checkout pull/1780` \
`$ git pull https://git.openjdk.org/jdk11u-dev pull/1780/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1780`

View PR using the GUI difftool: \
`$ git pr show -t 1780`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk11u-dev/pull/1780.diff">https://git.openjdk.org/jdk11u-dev/pull/1780.diff</a>

</details>

### CODE REVIEW

**Review by RealCLanger — APPROVED** [review]
LGTM

**Comment by RealCLanger** [comment]
can you please run GHA

**Comment by asgibbons** [comment]
Noob here - what is GHA?

**Comment by RealCLanger** [comment]
> Noob here - what is GHA?

GHA= GitHub Actions. You have to enable it in your [repository fork](https://github.com/asgibbons/jdk11u-dev/actions). Then you will also be able to trigger the run for your PR branch.

**Review by RealCLanger — CHANGES_REQUESTED** [review]
Hm after your merges the resulting change looks incorrect.

**Comment by asgibbons** [comment]
@RealCLanger - Sorry for the churn.  I had a slight mixup in the myriad of JDK trees I'm concurrently maintaining.  I fixed the issue and manual testing along with tier1 testing have successfully completed.

Please review and approve if this is acceptable.

**Comment by sviswa7** [comment]
@ascarpino Could you please help review this backport CR?

File: src/java.base/share/classes/com/sun/crypto/provider/CipherCore.java
Comment (RealCLanger) [inline src/java.base/share/classes/com/sun/crypto/provider/CipherCore.java:959]: Please add a space after `if`

**Review by RealCLanger — CHANGES_REQUESTED** [review]
I had a closer look now. Overall it seems ok. Could you maybe rename outWithPadding to internalOutput? That would bring the code closer to head and seems the more logical name for the buffer. Thanks.

File: src/java.base/share/classes/com/sun/crypto/provider/CipherCore.java
Comment (RealCLanger) [inline src/java.base/share/classes/com/sun/crypto/provider/CipherCore.java:971]: Here the variable should remain outWithPadding.

File: src/java.base/share/classes/com/sun/crypto/provider/CipherCore.java
Comment (RealCLanger) [inline src/java.base/share/classes/com/sun/crypto/provider/CipherCore.java:973]: Same here.

**Review by RealCLanger — CHANGES_REQUESTED** [review]

**Comment by RealCLanger** [comment]
Thanks. This looks good. I'll run it through SAP'S nightlies and will hopefully approve tomorrow when no regressions spotted.

**Comment by asgibbons** [comment]
Thank you very much @RealCLanger !

**Comment by ascarpino** [comment]
The backport looks consistent with what was put in recent releases.

**Comment by ascarpino** [comment]
Something that should be verified is that this doesn't expose failed auth tag decrypted GCM data or other GCM decryption.  There maybe some unexpected differences as this change was after GCM separated from CipherCore in 17.  I believe GCM still uses CipherCore in 11.

**Comment by asgibbons** [comment]
Thanks, @ascarpino.  Are there any tests I can run to verify that no GCM data have been exposed?  It appears to me that there's one place where decrypted data could be leaked (ShortBufferException), but this is the same in newer versions.

**Review by RealCLanger — APPROVED** [review]
Looks good from my perspective now. SAP tests passed. Can't say anything to the GCM topic brought up by @ascarpino, though.

**Comment by asgibbons** [comment]
@ascarpino I have been looking to see whether any data could be potentially leaked as a result of my change and cannot envision such a scenario.  It appears to me that my change is functionally identical to what it would be without the change.  Can you please provide more details?  Thanks.

**Comment by asgibbons** [comment]
I believe there is no additional risk to data leakage as the result of this change, so I'm proposing integration.  Thanks.

/integrate

**Comment by RealCLanger** [comment]
> I believe there is no additional risk to data leakage as the result of this change, so I'm proposing integration. Thanks.
> 
> /integrate

ok

/sponsor

**Pull request closed without merging**
